### PR TITLE
CORE-8460 UI Resiliency - Data Window

### PR DESCRIPTION
**Pull request by ash3rz**
This PR builds on top of PR #49 but with changes for the Data window.

Really the only tricky thing with this PR is some of the endpoints in the Disk Resource service facades go from terrain to the metadata service, not data-info.  I tried to check them all to make sure only the data-info endpoints got updated (don't want to disable the entire Data window if only the metadata service is down).

### CODE REVIEW

File: de-lib/src/main/java/org/iplantc/de/client/services/DiskResourceServiceFacade.java [hunk 1]
```diff
@@ -36,7 +38,7 @@
      * @param folder The parent of the subfolders to refresh from the service.
      * @param callback executed when RPC call completes.
      */
-    void refreshFolder(Folder folder, final AsyncCallback<List<Folder>> callback);
+    void refreshFolder(Folder folder, final DECallback<List<Folder>> callback);
```
Comment (ianmcorvidae) [inline de-lib/src/main/java/org/iplantc/de/client/services/DiskResourceServiceFacade.java]: Why is it these ones are `DECallback` rather than `DataCallback`? Couldn't quite tell what the difference was supposed to be (mostly looking through the PR to see if there's any tricky multi-service-endpoint edge cases I could rustle up)

File: de-lib/src/main/java/org/iplantc/de/client/services/DiskResourceServiceFacade.java [hunk 2]
```diff
@@ -129,7 +131,7 @@ void moveContents(final Folder sourceFolder,
      * @param dest id of the destination folder.
      * @param callback service success/failure callback
      */
-    void importFromUrl(String url, DiskResource dest, AsyncCallback<String> callback);
+    void importFromUrl(String url, DiskResource dest, DECallback<String> callback);
```
Comment (ianmcorvidae) [inline de-lib/src/main/java/org/iplantc/de/client/services/DiskResourceServiceFacade.java]: This one is + may turn into even more of a weird case; URL import right now goes from terrain to apps and doesn't use data-info. But, the plan is to have it go through terrain into data-info, which then calls out to apps (CORE-6214). I'm not sure how things go down in these cases.

File: de-lib/src/main/java/org/iplantc/de/client/services/DiskResourceServiceFacade.java (hunk 1 above)
Comment (ash3rz) [inline de-lib/src/main/java/org/iplantc/de/client/services/DiskResourceServiceFacade.java]: Mostly just to keep the service facades generic.  There probably won't be any cross-over with the disk resource / data window related callbacks, but for example, there could be a case where the Analysis window and the Apps window share the same callback and I want the service to be generic enough to handle either an `AppsCallback` or `AnalysisCallback` (in another PR coming up).

File: de-lib/src/main/java/org/iplantc/de/client/services/DiskResourceServiceFacade.java [hunk 3]
```diff
@@ -144,29 +146,28 @@ void moveContents(final Folder sourceFolder,
...
-    void deleteDiskResources(HasPaths diskResources, AsyncCallback<HasPaths> callback);
+    void deleteDiskResources(HasPaths diskResources, DECallback<HasPaths> callback);
 
     /**
      * @param resource the <code>DiskResource</code> for which metadata will be retrieved.
      * @param callback callback executed when service call completes.
      */
-    void getDiskResourceMetaData(DiskResource resource,
-                                 AsyncCallback<DiskResourceMetadataList> callback);
+    void getDiskResourceMetaData(DiskResource resource, DECallback<DiskResourceMetadataList> callback);
```
Comment (ianmcorvidae) [inline de-lib/src/main/java/org/iplantc/de/client/services/DiskResourceServiceFacade.java]: This and the set-metadata endpoint both go through terrain to data-info and then to metadata, which it sounds like you're aware of (but in any case, it's those two particularly)

File: de-lib/src/main/java/org/iplantc/de/client/services/DiskResourceServiceFacade.java (hunk 1 above)
Comment (ianmcorvidae) [inline de-lib/src/main/java/org/iplantc/de/client/services/DiskResourceServiceFacade.java]: Ah, I see. Since this is just a function parameter list it's better to use the more generic callback. With you now!

File: de-lib/src/main/java/org/iplantc/de/client/services/DiskResourceServiceFacade.java [hunk 4]
```diff
@@ -311,10 +302,10 @@ void setDiskResourceMetaData(DiskResource resource,
      * @param paths destination DR's path to which metadata will be copied.
      * @param callback callback object
      */
-            void
-            copyMetadata(final String srcUUID,
-                         final Splittable paths,
-                         final AsyncCallback<String> callback);
+
+    void copyMetadata(final String srcUUID,
```
Comment (ianmcorvidae) [inline de-lib/src/main/java/org/iplantc/de/client/services/DiskResourceServiceFacade.java]: (copy and save metadata both use the metadata service too, of course)

File: de-lib/src/main/java/org/iplantc/de/client/services/DiskResourceServiceFacade.java (hunk 2 above)
Comment (ash3rz) [inline de-lib/src/main/java/org/iplantc/de/client/services/DiskResourceServiceFacade.java]: Good catch - I'll update the presenter that calls this endpoint to `AppsCallback` since that makes the most sense right now.  We can switch it to `DataCallback` when CORE-6214 goes through.

File: de-lib/src/main/java/org/iplantc/de/client/services/DiskResourceServiceFacade.java (hunk 3 above)
Comment (ash3rz) [inline de-lib/src/main/java/org/iplantc/de/client/services/DiskResourceServiceFacade.java]: Yeah, I might need to do some more testing to make sure that if the metadata service is down it doesn't disable the Data window in some of these endpoints.  Hopefully that won't be the case.

Comment (psarando) [inline de-lib/src/main/java/org/iplantc/de/client/services/DiskResourceServiceFacade.java]: I'm not sure if the UI can tell which service is down if it gets a 503 from this call.
In my previous comment I suggest that maybe the UI should assume data-info is down if this call returns a 503, and if it's just metadata, then the `retry` button will re-enable the data window right away?

File: de-lib/src/main/java/org/iplantc/de/client/services/DiskResourceServiceFacade.java (hunk 2 above)
Comment (ianmcorvidae) [inline de-lib/src/main/java/org/iplantc/de/client/services/DiskResourceServiceFacade.java]: That's a good idea; we should be able to make the nginx return that automatically if we add a JSON body to the 502/503/504 response, and then we just need to make sure our error-handling code passes that info along. That'll probably need more investigation, and it's possible some of these endpoints should more gracefully respond when stuff is down anyway (e.g. the manifest endpoint)

File: de-lib/src/main/java/org/iplantc/de/client/services/FileEditorServiceFacade.java [hunk 5]
```diff
@@ -17,7 +18,7 @@
      *  @param file desired manifest's file ID (path).
      * @param callback executes when RPC call is complete.
      */
-    void getManifest(File file, AsyncCallback<Manifest> callback);
+    void getManifest(File file, DECallback<Manifest> callback);
```
Comment (ianmcorvidae) [inline de-lib/src/main/java/org/iplantc/de/client/services/FileEditorServiceFacade.java]: Here's another one to test an edge case: the manifest is through data-info, but it uses the tree-urls service in the background. Most likely we should make a fallback in data-info that causes this to not throw the 503 the whole way upward (and tree-urls is rarely down, it's so tiny), but for now it's something to check.

File: de-lib/src/main/java/org/iplantc/de/client/services/DiskResourceServiceFacade.java (hunk 3 above)
Comment (ash3rz) [inline de-lib/src/main/java/org/iplantc/de/client/services/DiskResourceServiceFacade.java]: Hm... right now the UI is assuming there's a link between windows and services.  So the presenters decide which type of callback, in this case `DataCallback`, to use.  If that callback fails, then DECallback knows which windows to disable.  So if we have a callback that can return a 502/503 but it shouldn't disable the Data window, then it shouldn't be a `DataCallback` type. 

The `retry` button will only re-enable a window if the call that initially failed begins to succeed, because it only resends that first failing call.

File: de-lib/src/main/java/org/iplantc/de/client/services/FileEditorServiceFacade.java (hunk 5 above)
Comment (ash3rz) [inline de-lib/src/main/java/org/iplantc/de/client/services/FileEditorServiceFacade.java]: From what I can tell, if I enable maintenance on tree-urls, I can still view a file (which calls the manifest endpoint) and the Data window stays enabled.  So we're good here 👍

File: de-lib/src/main/java/org/iplantc/de/client/services/DiskResourceServiceFacade.java (hunk 3 above)
Comment (ash3rz) [inline de-lib/src/main/java/org/iplantc/de/client/services/DiskResourceServiceFacade.java]: So it seems like getting metadata currently won't disable the Data window, but trying to copy/save metadata will. 😖

Comment (ash3rz) [inline de-lib/src/main/java/org/iplantc/de/client/services/DiskResourceServiceFacade.java]: (and you get a ton of 503 error pop ups from the Data window's Details panel, but that's a different story)

Comment (ash3rz) [inline de-lib/src/main/java/org/iplantc/de/client/services/DiskResourceServiceFacade.java]: I think for now, I'll switch the get and set metadata endpoints back to the old `AsyncCallback` so they can't disable the Data window.

**Comment by ash3rz** [comment]
Thanks for the review!
